### DISCUSSION TITLE
Passing a React Server Function as Props to a Client Component in Next.js - Is This a Valid Approach?

### DISCUSSION THREAD

**Discussion by theishu4** [vercel/next.js#72270] [Help] [answered]
### Summary

I'm trying to pass a server function (previously called a "server action") to a client component, specifically a `<Button />` component in my React app. Here’s my code:

### Code Example

```javascript
import { Button } from "@/components/ui/button";
import { auth } from "@/auth";


const Navbar = async () => {
  const session = await auth();

  async function signOutAction() {
    "use server";
    await signOut({ redirectTo: "/" });
  }

  return (
    <div>
      {session ? (
        // 👇 Is this a valid and preferred way of doing it?
        // The code works as expected without errors or warnings.
        <Button onClick={signOutAction}>Logout</Button>
      ) : null}
    </div>
  );
};
```

The code works without any warnings or errors, but I'm unsure if this is the best approach. Specifically, is it safe or recommended to pass the `signOutAction` server function as the `onClick` prop directly to the `<Button />` component?

### Alternative Approach

If the above approach is not recommended, would the following alternative be better?

```javascript
const Navbar = async () => {
  const session = await auth();

  async function signOutAction() {
    "use server";
    await signOut({ redirectTo: "/" });
  }

  return (
    <div>
      {session ? (
        // 👇 Is this approach better or more standard?
        <form action={signOutAction}>
          <Button type="submit">Logout</Button>
        </form>
      ) : null}
    </div>
  );
};
```

### Question

1. Is it acceptable to pass a server function as a prop directly to a client component?
2. If not, is wrapping the `<Button />` in a `<form>` and using `action={signOutAction}` the preferred way to handle server actions for client components?

Any guidance on the correct approach or best practices for handling server actions in this context would be greatly appreciated!

**Comment by nirvikpurkait** [comment, marked as answer]
You can pass server-action to a client component (button) and that is fine. Here you have a use case of using form, as you are logging out, but what if you have some other scenario, i.e: toggle a task between complete and incomplete, you wouldn't want to create a form element only for that. The documentation also suggest, you can pass server-action as prop. [see here](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#passing-actions-as-props).

If this is helpful, mark as answered.

**Comment by icyJoseph** [comment]
This is fine, literally documented here, https://react.dev/reference/rsc/server-actions#creating-a-server-action-from-a-server-component
